src/main/scala/zio/s3/Live.scala

```diff
@@ -105,78 +108,78 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
-                          _.uploadPart(
-                            UploadPartRequest
-                              .builder()
-                              .bucket(bucketName)
-                              .key(key)
-                              .partNumber(partNumber.toInt + 1)
-                              .uploadId(uploadId)
-                              .contentLength(chunk.length.toLong)
-                              .build(),
-                            AsyncRequestBody.fromBytes(chunk.toArray)
-                          )
-                        ).map(r => CompletedPart.builder().partNumber(partNumber.toInt + 1).eTag(r.eTag()).build())
-                    }
-                    .runCollect
-                    .mapError(S3ExceptionUtils.fromThrowable)
-
-      _        <- execute(
-                    _.completeMultipartUpload(
-                      CompleteMultipartUploadRequest
-                        .builder()
-                        .bucket(bucketName)
-                        .key(key)
-                        .multipartUpload(CompletedMultipartUpload.builder().parts(parts.asJavaCollection).build())
-                        .uploadId(uploadId)
-                        .build()
-                    )
-                  )
+      uploadId      <- execute(
+                         _.createMultipartUpload(
+                           UploadOptions
+                             .multipartUploadBuilder(options)
+                             .bucket(bucketName)
+                             .key(key)
+                             .build()
+                         )
+                       ).map(_.uploadId())
+
+      parts         <- content
+                         .chunkN(partSize)
+                         .mapChunks(Chunk.single)
+                         .zipWithIndex
+                         .mapMPar(parallelism) {
+                           case (chunk, partNumber) => uploadPart(bucketName, key, partNumber.toInt + 1, uploadId, chunk)
+                         }
+                         .runCollect
+                         .mapError(S3ExceptionUtils.fromThrowable)
+
+      nonEmptyParts <- if (parts.nonEmpty) UIO(parts)
+                       else uploadPart(bucketName, key, 1, uploadId, Chunk.empty).map(List(_))
+
+      _             <- execute(
+                         _.completeMultipartUpload(
+                           CompleteMultipartUploadRequest
+                             .builder()
+                             .bucket(bucketName)
+                             .key(key)
+                             .multipartUpload(CompletedMultipartUpload.builder().parts(nonEmptyParts.asJavaCollection).build())
+                             .uploadId(uploadId)
+                             .build()
+                         )
+                       )
     } yield ()
 
+  private def uploadPart(bucketName: String, key: String, partNumber: Int, uploadId: String, chunk: Chunk[Byte]) =
```

Comment:
from my solution you dont need to exrtact this method

Reply:
https://github.com/zio/zio-s3/pull/117/commits/d6b2f1b0bb06dfc48d245b6409ab4654c6479d82